mri/reconstructors/tests/test_reconstructors.py

```diff
@@ -0,0 +1,249 @@
+            # For Undecimated wavelets, the wavelet_name is wavelet_id
+            linear_op = WaveletUD2(
+                wavelet_id=wavelet_name,
+                nb_scale=nb_scale,
+                n_coils=n_coils,
+                n_jobs=n_jobs,
+                verbose=verbose,
+            )
+        return linear_op
+
+    def test_single_channel_reconstruction(self):
+        """ Test all the registered transformations for
+        single channel reconstructor.
+        """
+        print("Process test for SingleChannelReconstructor ::")
+        for i in range(len(self.test_cases)):
+            print("Test Case " + str(i) + " " + str(self.test_cases[i]))
+            image, nb_scale, optimizer, recon_type, name = self.test_cases[i]
+            if optimizer == 'condatvu':
+                formulation = "analysis"
+            else:
+                formulation = "synthesis"
+            if recon_type == 'cartesian':
+                fourier = FFT(
+                    samples=convert_mask_to_locations(self.mask),
+                    shape=image.shape)
+            else:
+                fourier = NonCartesianFFT(
+                    samples=convert_mask_to_locations(self.mask),
+                    shape=image.shape)
+            kspace_data = fourier.op(image.data)
+            linear_op = self.get_linear_operator(
+                wavelet_name=name,
+                dimension=len(fourier.shape),
+                nb_scale=3,
+            )
+            reconstructor = SingleChannelReconstructor(
+                fourier_op=fourier,
+                linear_op=linear_op,
+                gradient_method=formulation,
+                verbose=0,
+            )
+            x_final, costs, _ = reconstructor.reconstruct(
+                kspace_data=kspace_data,
+                optimization_alg=optimizer,
+                num_iterations=self.num_iter,
+            )
+            fourier_0 = FFT(
+                samples=convert_mask_to_locations(self.mask),
+                shape=image.shape,
+            )
+            data_0 = fourier_0.op(image.data)
+            # mu is 0 for above single channel reconstruction and
+            # hence we expect the result to be the inverse fourier transform
+            np.testing.assert_allclose(
+                x_final, fourier_0.adj_op(data_0))
+
+    def test_self_calibrating_reconstruction(self):
+        """ Test all the registered transformations.
+        """
+        self.num_channels = 2
+        num_iter = 2
+        print("Process test for SelfCalibratingReconstructor ::")
+        for i in range(len(self.test_cases)):
+            print("Test Case " + str(i) + " " + str(self.test_cases[i]))
+            image, nb_scale, optimizer, recon_type, name = self.test_cases[i]
+            if recon_type == 'cartesian':
+                # TODO fix SelfCalibrating recon for cartesian case
+                print("Skipping Test case as SelfCalibrationReconstructor is "
```

Comment:
I'm sorry but who told you that ?

Reply:
I am not saying about the model itself, I mean our codes needs to be adapted for this. This will become a separate issue to fix.
The Smap estimation fails for cartesian case, as the data provided needs to be moved using convert_mask_to_locations. 
I understand that this is a small task, but I didnt want to bloat up this PR